pkg/daemon/ceph/osd/kms/kms.go

```diff
@@ -145,6 +147,17 @@ func (c *Config) PutSecret(secretName, secretValue string) error {
 		}
 	}
 
+	if c.IsAzure() {
+		v, err := InitAzure(c.ClusterInfo.Context, c.context, c.ClusterInfo.Namespace, c.clusterSpec.Security.KeyManagementService.ConnectionDetails)
+		if err != nil {
+			return errors.Wrap(err, "failed to init azure key vault")
+		}
+		err = put(v, GenerateOSDEncryptionSecretName(secretName), secretValue, map[string]string{})
+		if err != nil {
+			return errors.Wrap(err, "failed to put secret in azure key vault")
+		}
+	}
+
```

Comment:
I see 3 instances of this same code block. What is the reason for that?

Reply:
They are for `put`, `get` and delete` operations.

Reply:
Got it. Thanks for explaining.

In general, I see this as a code smell. Methods like put/get shouldn't be doing KMS type checking. I think this is highlighting that our KMS support was hastily implemented and has been clunkily extended over time. I don't think this needs to be resolved in this PR, but it would be good to consider cleaning up the KMS code afterwards.

The better implementation would be to use a Golang interface that defines all the KMS methods, and then have each KMS implement all the methods. That would allow us to implement new KMSes more easily, it would make unit testing easier, and it would allow us to do functional e2e testing of KMSes without having to do a full install of Rook.